Add database versioning

dospuntos opened the pull request:
This is an attempt to prepare for possible future DB changes (e.g. when using system locale for dates and currencies) without the need for a conversion tool. Users upgrading from previous versions (or even before versioning was implemented) will automatically have any DB changes applied.

I'm not sure if my logic is correct or if there is a better way to implement this, any feedback appreciated. It seems to work with databases from CapitalBe 1.1.0 from HaikuDepot. Maybe an automatic backup of the current database would be a good idea in order to protect user data.

Each time a release requires database changes, the version number in `Database.cpp` is incremented, and the required logic (changing tables and migrating data) must be added as a new if-clause in `Database::ApplyMigrations()`. This function handles each version incrementally.

dospuntos commented:
I updated the code to make a backup of the database before the migrations are applied, in case something goes wrong.

The database is now being changed (some unused columns removed from the locale settings), and an upgraded database is not compatible with the original CapitalBe (from HaikuDepot). More columns can be removed when the system currency format is implemented. Replacing `MyAccountData` with the backup file `MyAccountData_backup_0` makes it possible to open the original app again.

> localize the alert text...

It looks like `ShowAlert()` and `ShowBug()` crashes de app when called from `Database.cpp`, is there another way to show error messages? Right now it's a mix of `printf()`, `std::cerr` etc.

dospuntos commented:
As far as I've been able to test, this now works, but I might have missed some edge cases.

To test:
- Install and run CapitalBe 1.1.0 from HaikuDepot
- Add a few accounts and transactions
- Close CapitalBe 1.1.0
- Build and run *this* version

All accounts and transactions should show up, with a few things to note:

- All accounts are set to custom currency (but will have the original currency symbol)
- A backup of the original database has been saved to `/boot/home/config/settings/CapitalBe/MyAccountData_backup_0`
- Re-running CapitalBe 1.1.0 will not work (database is not backwards compatible), but the backup can be renamed to `MyAccountData` if needed.

humdingerb commented:
First off, running the old CapitalBe makes our improvements quite stunning! :)

The old CapitalBe always crashes on startup, if you have scheduled transactions. Without those, the new CapitalBe seems to import the old data correctly.
With scheduled transactions, though, I get this crash of the new CapitalBe: 
[CapitalBe-4113-debug-18-06-2024-15-04-44.report.txt](https://github.com/user-attachments/files/15888604/CapitalBe-4113-debug-18-06-2024-15-04-44.report.txt)

Here's the old CapitalBe's backed-up data (renamed to .txt to be able to upload here: 
[MyAccountData_backup_0.txt](https://github.com/user-attachments/files/15888649/MyAccountData_backup_0.txt)

dospuntos commented:
Thanks for testing and sending your database. I wonder how many users the original version has, if any, with scheduled transactions, but either way it should be fixed. I'll see if I can figure it out.

And yes, the the new version has a lot of improvements, both with stability and GUI. You have a great eye for improvements on the user experience!

dospuntos commented:
The problem was that the migration script inserts a new table column for scheduled transactions (for a planned fix for #76), but the code wasn't aware of the change. I've added a temporary fix, and the import should work now. 

I also noticed from the backup you attached that it was taken *after* the `db_version` table was created, but before the actual migration. I've updated the code to create the backup before any changes are made.

humdingerb commented:
OK, seems to be working now! Good you're doing the coding for the backend and logic and all that, because I'm probably terrible at that... :)

Ready to merge?

dospuntos commented:
Just a heads up that if there are changes to the migration script in a later PR (e.g. to handle scheduled transfers between accounts or other improvements), the database might not be compatible with builds from other branches or previous pull requests. It will of course be compatible with the old CapitalBe release.

If that's OK then please merge.

(Just thought I'd mention it, although I think there's already been cases where different builds expect different database structure.)

dospuntos commented:
Updated PR with more DB migration functions. The migration now also deescapes all database entries, like payee, category and memos. Function calls to escape and deescape strings have been removed from the code.

This should update a database from the old CapitalBe to the current version, keeping all data.

Attached is a database with some test data, compatible with the old CapitalBe 1.1.0. There are a few accounts, transactions with random names, categories and budget entries (no scheduled transactions, though). Added .txt extension to be able to upload the file to GitHub.

[MyAccountData.txt](https://github.com/user-attachments/files/15963623/MyAccountData.txt)

humdingerb commented:
I gave it a spin and your test data and my (very small) test case appears to be imported fine.
Let me know if you deem it ready to merge.

dospuntos commented:
Thanks for testing, I think this can be merged now.

humdingerb commented:
Nice work!
